state.go

```diff
@@ -317,6 +325,38 @@ func (m *Memberlist) probeNode(node *nodeState) {
 	m.suspectNode(&s)
 }
 
+// Ping initiates a ping to the node with the specified name.
+func (m *Memberlist) Ping(node string, addr net.IP, port uint16) (time.Duration, error) {
+	// Prepare a ping message and setup an ack handler.
+	ping := ping{SeqNo: m.nextSeqNo(), Node: node}
+	ackCh := make(chan ackMessage, m.config.IndirectChecks+1)
+	m.setAckChannel(ping.SeqNo, ackCh, m.config.ProbeInterval)
+
+	// Send a ping to the node.
+	destAddr := &net.UDPAddr{IP: addr, Port: int(port)}
+	if err := m.encodeAndSendMsg(destAddr, pingMsg, &ping); err != nil {
+		return 0, err
+	}
+
+	// Mark the sent time here, which should be after any pre-processing and
+	// system calls to do the actual send. This probably under-reports a bit,
+	// but it's the best we can do.
+	sent := time.Now()
+
+	// Wait for response or timeout.
+	select {
+	case v := <-ackCh:
+		if v.Complete == true {
+			return v.Timestamp.Sub(sent), nil
```

Comment:
It's actually the two-way timestamp, though the plumbing of that is pretty convoluted. The timestamp comes from here https://github.com/hashicorp/memberlist/blob/master/net.go#L295 and gets passed down through function calls and then stuck in the channel for this to read out. This tries to get the timestamp up as close to the IO as possible.